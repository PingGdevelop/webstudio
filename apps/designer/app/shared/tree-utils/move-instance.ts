import { type Instance } from "@webstudio-is/react-sdk";
import { insertInstanceMutable } from "./insert-instance";

type InstanceMoveSpec = {
  id: Instance["id"];
  newParentId: Instance["id"];
  newIndex: number;
};

const extractInstance = (
  treeNode: Instance,
  id: Instance["id"]
): Instance | null => {
  const index = treeNode.children.findIndex(
    (child) => typeof child !== "string" && child.id === id
  );

  if (index !== -1) {
    const instance = treeNode.children[index];

    // for TypeScript
    if (typeof instance === "string") {
      throw new Error("Cannot happen");
    }

    treeNode.children.splice(index, 1);
    return instance;
  }

  for (const child of treeNode.children) {
    if (typeof child !== "string") {
      const extracted = extractInstance(child, id);
      if (extracted !== null) {
        return extracted;
      }
    }
  }

  return null;
};

export const moveInstanceMutable = (
  treeRoot: Instance,
  spec: InstanceMoveSpec
) => {
  const instance = extractInstance(treeRoot, spec.id);

  // mostly for TypeScript
  if (instance === null) {
    throw new Error("Cannot move instance that doesn't exist");
  }

  insertInstanceMutable(treeRoot, instance, {
    parentId: spec.newParentId,
    position: spec.newIndex,
  });

  return treeRoot;
};
