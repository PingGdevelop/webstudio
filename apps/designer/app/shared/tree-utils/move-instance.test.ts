import { moveInstanceMutable } from "./move-instance";
import { createInstance } from "./create-instance";

describe("Insert instance", () => {
  test(`No changes`, () => {
    const tree = createInstance({
      id: "root",
      component: "Box",
      children: [
        createInstance({ id: "1", component: "Box" }),
        createInstance({ id: "2", component: "Box" }),
      ],
    });

    moveInstanceMutable(tree, { id: "1", newParentId: "root", newIndex: 0 });

    expect(tree).toMatchSnapshot();
  });

  test(`Change index`, () => {
    const tree = createInstance({
      id: "root",
      component: "Box",
      children: [
        createInstance({ id: "1", component: "Box" }),
        createInstance({ id: "2", component: "Box" }),
      ],
    });

    moveInstanceMutable(tree, { id: "1", newParentId: "root", newIndex: 1 });

    expect(tree).toMatchSnapshot();
  });

  test(`Change index nested`, () => {
    const tree = createInstance({
      id: "root",
      component: "Box",
      children: [
        createInstance({
          id: "1",
          component: "Box",
          children: [
            createInstance({ id: "3", component: "Box" }),
            createInstance({ id: "4", component: "Box" }),
          ],
        }),
        createInstance({ id: "2", component: "Box" }),
      ],
    });

    moveInstanceMutable(tree, { id: "3", newParentId: "1", newIndex: 1 });

    expect(tree).toMatchSnapshot();
  });

  test(`Change parent`, () => {
    const tree = createInstance({
      id: "root",
      component: "Box",
      children: [
        createInstance({
          id: "1",
          component: "Box",
          children: [
            createInstance({ id: "3", component: "Box" }),
            createInstance({ id: "4", component: "Box" }),
          ],
        }),
        createInstance({ id: "2", component: "Box" }),
      ],
    });

    moveInstanceMutable(tree, { id: "3", newParentId: "2", newIndex: 0 });

    expect(tree).toMatchSnapshot();
  });
});
