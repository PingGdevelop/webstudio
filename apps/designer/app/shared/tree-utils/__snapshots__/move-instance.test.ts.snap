// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Insert instance Change index 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "2",
    },
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "1",
    },
  ],
  "component": "Box",
  "cssRules": Array [],
  "id": "root",
}
`;

exports[`Insert instance Change index nested 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "component": "Box",
          "cssRules": Array [],
          "id": "4",
        },
        Object {
          "children": Array [],
          "component": "Box",
          "cssRules": Array [],
          "id": "3",
        },
      ],
      "component": "Box",
      "cssRules": Array [],
      "id": "1",
    },
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "2",
    },
  ],
  "component": "Box",
  "cssRules": Array [],
  "id": "root",
}
`;

exports[`Insert instance Change parent 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "component": "Box",
          "cssRules": Array [],
          "id": "4",
        },
      ],
      "component": "Box",
      "cssRules": Array [],
      "id": "1",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "component": "Box",
          "cssRules": Array [],
          "id": "3",
        },
      ],
      "component": "Box",
      "cssRules": Array [],
      "id": "2",
    },
  ],
  "component": "Box",
  "cssRules": Array [],
  "id": "root",
}
`;

exports[`Insert instance No changes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "1",
    },
    Object {
      "children": Array [],
      "component": "Box",
      "cssRules": Array [],
      "id": "2",
    },
  ],
  "component": "Box",
  "cssRules": Array [],
  "id": "root",
}
`;
